// devexpress-reporting/designer/dataObjects/parameters/parameterTypesHelper.js
/**
* DevExpress HTML/JS Reporting (designer\dataObjects\parameters\parameterTypesHelper.js)
* Version:  24.1.7
* Build date: Oct 28, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { enumValueConverter, floatValueConverter, integerValueConverter, parseDate } from '@devexpress/analytics-core/analytics-internal';
import { validateGuid } from '@devexpress/analytics-core/analytics-widgets-internal';
class ParameterTypesHelper {
    _getTypeInfo(typeName) {
        const values = ParameterTypesHelper.typeValues.filter((type) => { return type.value === typeName; });
        if (values.length > 0)
            return values[0];
        if (this.enumValueTypes?.length) {
            const enumValues = this.enumValueTypes.filter((enumInfo) => enumInfo.value === typeName);
            if (enumValues.length > 0)
                return enumValues[0];
        }
        return null;
    }
    _tryConvertValue(value, typeName) {
        const condition = val => { return val !== void 0 && val !== null && !isNaN(typeof val === 'string' ? '' : val); };
        if (!condition(value)) {
            return { isValid: false, newValue: null };
        }
        const typeValue = this._getTypeInfo(typeName), newValue = (typeValue && typeValue.valueConverter) ? typeValue.valueConverter(value, this.getDefaultValue(typeName)) : value;
        return { isValid: condition(newValue), newValue: newValue };
    }
    convertSingleValue(value, typeName) {
        const result = this._tryConvertValue(value, typeName);
        return result.isValid ? result.newValue : this.getDefaultValue(typeName);
    }
    getSpecifics(typeName) {
        const typeValue = this._getTypeInfo(typeName);
        return typeValue ? typeValue.specifics : 'default';
    }
    getIcon(typeName) {
        const typeValue = this._getTypeInfo(typeName);
        return typeValue && typeValue.icon;
    }
    getDefaultValue(typeName) {
        const typeValue = this._getTypeInfo(typeName);
        let _value = typeValue ? typeValue.defaultValue : '';
        if (_value === '' && !!this.knownEnums?.length) {
            const enumType = this.knownEnums.filter((enumInfo) => enumInfo.enumType === typeName)[0];
            if (!!enumType) {
                _value = enumType.values[0]?.value ?? 0;
            }
        }
        else if (_value instanceof Date) {
            _value = new Date(_value);
        }
        return _value;
    }
    getEnumTypeValues() {
        if (!!this.knownEnums?.length) {
            const enumtypeValues = [];
            this.knownEnums.forEach(knownEnum => {
                enumtypeValues.push({
                    value: knownEnum.enumType,
                    displayValue: knownEnum.enumType,
                    defaultValue: knownEnum.values[0].value,
                    specifics: 'Enum',
                    valueConverter: function (val, defaultValue) { return enumValueConverter(val, defaultValue, knownEnum.values); }
                });
            });
            return enumtypeValues;
        }
    }
    constructor(knownEnums) {
        this.knownEnums = knownEnums;
        this.enumValueTypes = this.getEnumTypeValues();
    }
}
ParameterTypesHelper.defaultGuidValue = '00000000-0000-0000-0000-000000000000';
ParameterTypesHelper.typeValues = [
    { value: 'System.String', displayValue: 'String', defaultValue: '', specifics: 'String', valueConverter: function (val) { return val.toString(); }, localizationId: 'UtilsUIStringId.Parameter_Type_String' },
    { value: 'System.DateTime', displayValue: 'Date', defaultValue: new Date(new Date().setHours(0, 0, 0, 0)), specifics: 'Date', valueConverter: function (val) { return parseDate(val); }, localizationId: 'UtilsUIStringId.Parameter_Type_DateTime' },
    { value: 'System.Int16', displayValue: 'Number (16 bit integer)', defaultValue: '0', specifics: 'Integer', valueConverter: function (val, defaultValue) { return integerValueConverter(val, defaultValue, 'System.Int16'); }, localizationId: 'UtilsUIStringId.Parameter_Type_Int16' },
    { value: 'System.Int32', displayValue: 'Number (32 bit integer)', defaultValue: '0', specifics: 'Integer', valueConverter: function (val, defaultValue) { return integerValueConverter(val, defaultValue, 'System.Int32'); }, localizationId: 'UtilsUIStringId.Parameter_Type_Int32' },
    { value: 'System.Int64', displayValue: 'Number (64 bit integer)', defaultValue: '0', specifics: 'Integer', valueConverter: function (val, defaultValue) { return integerValueConverter(val, defaultValue, 'System.Int64'); }, localizationId: 'UtilsUIStringId.Parameter_Type_Int64' },
    { value: 'System.Single', displayValue: 'Number (floating-point)', defaultValue: '0', specifics: 'Float', valueConverter: function (val, defaultValue) { return floatValueConverter(val, defaultValue, 'System.Single'); }, localizationId: 'UtilsUIStringId.Parameter_Type_Float' },
    { value: 'System.Double', displayValue: 'Number (double-precision floating-point)', defaultValue: '0', specifics: 'Float', valueConverter: function (val, defaultValue) { return floatValueConverter(val, defaultValue, 'System.Double'); }, localizationId: 'UtilsUIStringId.Parameter_Type_Double' },
    { value: 'System.Decimal', displayValue: 'Number (decimal)', defaultValue: '0', specifics: 'Float', valueConverter: function (val, defaultValue) { return floatValueConverter(val, defaultValue, 'System.Decimal'); }, localizationId: 'UtilsUIStringId.Parameter_Type_Decimal' },
    { value: 'System.Boolean', displayValue: 'Boolean', defaultValue: false, specifics: 'Bool', valueConverter: function (val) { return String(val).toLowerCase() === 'true' ? true : (String(val).toLowerCase() === 'false' ? false : null); }, localizationId: 'UtilsUIStringId.Parameter_Type_Boolean' },
    { value: 'System.Guid', displayValue: 'Guid', defaultValue: ParameterTypesHelper.defaultGuidValue, valueConverter: function (val) { return validateGuid(val) ? val : ParameterTypesHelper.defaultGuidValue; }, specifics: 'guid', localizationId: 'UtilsUIStringId.Parameter_Type_Guid' }
];
export { ParameterTypesHelper };
