// @devexpress/analytics-core/query-builder/dataSource/sql/sqlDataConnection.js
/**
* DevExpress Analytics (query-builder\dataSource\sql\sqlDataConnection.js)
* Version:  24.1.7
* Build date: Oct 28, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { ModelSerializer } from '../../../serializer/serializer';
import { parseBool } from '../../../core/utils/parsers';
export class SqlDataConnection {
    static from(model, serializer) {
        return new SqlDataConnection(model, serializer);
    }
    static toJson(value, serializer, refs) {
        return serializer.serialize(value, sqlDataConnectionSerializationInfo, refs);
    }
    getInfo() {
        return sqlDataConnectionSerializationInfo;
    }
    constructor(model, serializer) {
        this.name = ko.observable();
        this.parameteres = ko.observable();
        this.fromAppConfig = ko.observable(true);
        serializer = serializer || new ModelSerializer();
        serializer.deserialize(this, model);
    }
}
const sqlDataConnectionSerializationInfo = [
    { propertyName: 'name', modelName: '@Name' },
    { propertyName: 'parameteres', modelName: 'Parameters' },
    { propertyName: 'fromAppConfig', modelName: '@FromAppConfig', defaultVal: false, from: parseBool }
];
