// @devexpress/analytics-core/accessibility/_controlElementBase.js
/**
* DevExpress Analytics (accessibility\_controlElementBase.js)
* Version:  24.1.7
* Build date: Oct 28, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { Disposable } from '../serializer/disposable';
import { addDisposeCallback } from '../serializer/_internal';
export class AccessibilityControlElementBase extends Disposable {
    dispose() {
        super.dispose();
        this._eventListeners.forEach((item) => { item.element.removeEventListener(item.eventType, item.listener); });
        this._eventListeners.length = 0;
        this.element = null;
    }
    constructor(element) {
        super();
        this.element = element;
        this._eventListeners = [];
    }
    addListener(element, eventType, handler) {
        const listener = (e) => {
            handler.call(this, e);
        };
        element.addEventListener(eventType, listener);
        addDisposeCallback(element, function () {
            element.removeEventListener(eventType, listener);
        });
        this._eventListeners.push({ element: element, eventType: eventType, listener: listener });
    }
    setTabIndex(index) {
        this.element.setAttribute('tabindex', index);
    }
    setFocus() {
        this.element.focus();
    }
}
