// @devexpress/analytics-core/widgets/criteria/operators/options/binary.js
/**
* DevExpress Analytics (widgets\criteria\operators\options\binary.js)
* Version:  24.1.7
* Build date: Oct 28, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
export var BinaryOperatorType;
(function (BinaryOperatorType) {
    BinaryOperatorType[BinaryOperatorType["Equal"] = 0] = "Equal";
    BinaryOperatorType[BinaryOperatorType["NotEqual"] = 1] = "NotEqual";
    BinaryOperatorType[BinaryOperatorType["Greater"] = 2] = "Greater";
    BinaryOperatorType[BinaryOperatorType["Less"] = 3] = "Less";
    BinaryOperatorType[BinaryOperatorType["LessOrEqual"] = 4] = "LessOrEqual";
    BinaryOperatorType[BinaryOperatorType["GreaterOrEqual"] = 5] = "GreaterOrEqual";
    BinaryOperatorType[BinaryOperatorType["Like"] = 6] = "Like";
    BinaryOperatorType[BinaryOperatorType["BitwiseAnd"] = 7] = "BitwiseAnd";
    BinaryOperatorType[BinaryOperatorType["BitwiseOr"] = 8] = "BitwiseOr";
    BinaryOperatorType[BinaryOperatorType["BitwiseXor"] = 9] = "BitwiseXor";
    BinaryOperatorType[BinaryOperatorType["Divide"] = 10] = "Divide";
    BinaryOperatorType[BinaryOperatorType["Modulo"] = 11] = "Modulo";
    BinaryOperatorType[BinaryOperatorType["Multiply"] = 12] = "Multiply";
    BinaryOperatorType[BinaryOperatorType["Plus"] = 13] = "Plus";
    BinaryOperatorType[BinaryOperatorType["Minus"] = 14] = "Minus";
})(BinaryOperatorType || (BinaryOperatorType = {}));
