// @devexpress/analytics-core/widgets/criteria/operators/options/function.js
/**
* DevExpress Analytics (widgets\criteria\operators\options\function.js)
* Version:  24.1.7
* Build date: Oct 28, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
export var FunctionOperatorType;
(function (FunctionOperatorType) {
    FunctionOperatorType[FunctionOperatorType["None"] = 0] = "None";
    FunctionOperatorType[FunctionOperatorType["Custom"] = 1] = "Custom";
    FunctionOperatorType[FunctionOperatorType["CustomNonDeterministic"] = 2] = "CustomNonDeterministic";
    FunctionOperatorType[FunctionOperatorType["Iif"] = 3] = "Iif";
    FunctionOperatorType[FunctionOperatorType["InRange"] = 4] = "InRange";
    FunctionOperatorType[FunctionOperatorType["IsNull"] = 5] = "IsNull";
    FunctionOperatorType[FunctionOperatorType["IsNullOrEmpty"] = 6] = "IsNullOrEmpty";
    FunctionOperatorType[FunctionOperatorType["Trim"] = 7] = "Trim";
    FunctionOperatorType[FunctionOperatorType["Len"] = 8] = "Len";
    FunctionOperatorType[FunctionOperatorType["Substring"] = 9] = "Substring";
    FunctionOperatorType[FunctionOperatorType["Upper"] = 10] = "Upper";
    FunctionOperatorType[FunctionOperatorType["Lower"] = 11] = "Lower";
    FunctionOperatorType[FunctionOperatorType["Concat"] = 12] = "Concat";
    FunctionOperatorType[FunctionOperatorType["Ascii"] = 13] = "Ascii";
    FunctionOperatorType[FunctionOperatorType["Char"] = 14] = "Char";
    FunctionOperatorType[FunctionOperatorType["ToStr"] = 15] = "ToStr";
    FunctionOperatorType[FunctionOperatorType["Replace"] = 16] = "Replace";
    FunctionOperatorType[FunctionOperatorType["Reverse"] = 17] = "Reverse";
    FunctionOperatorType[FunctionOperatorType["Insert"] = 18] = "Insert";
    FunctionOperatorType[FunctionOperatorType["CharIndex"] = 19] = "CharIndex";
    FunctionOperatorType[FunctionOperatorType["Remove"] = 20] = "Remove";
    FunctionOperatorType[FunctionOperatorType["Abs"] = 21] = "Abs";
    FunctionOperatorType[FunctionOperatorType["Sqr"] = 22] = "Sqr";
    FunctionOperatorType[FunctionOperatorType["Cos"] = 23] = "Cos";
    FunctionOperatorType[FunctionOperatorType["Sin"] = 24] = "Sin";
    FunctionOperatorType[FunctionOperatorType["Atn"] = 25] = "Atn";
    FunctionOperatorType[FunctionOperatorType["Exp"] = 26] = "Exp";
    FunctionOperatorType[FunctionOperatorType["Log"] = 27] = "Log";
    FunctionOperatorType[FunctionOperatorType["Rnd"] = 28] = "Rnd";
    FunctionOperatorType[FunctionOperatorType["Tan"] = 29] = "Tan";
    FunctionOperatorType[FunctionOperatorType["Power"] = 30] = "Power";
    FunctionOperatorType[FunctionOperatorType["Sign"] = 31] = "Sign";
    FunctionOperatorType[FunctionOperatorType["Round"] = 32] = "Round";
    FunctionOperatorType[FunctionOperatorType["Ceiling"] = 33] = "Ceiling";
    FunctionOperatorType[FunctionOperatorType["Floor"] = 34] = "Floor";
    FunctionOperatorType[FunctionOperatorType["Max"] = 35] = "Max";
    FunctionOperatorType[FunctionOperatorType["Min"] = 36] = "Min";
    FunctionOperatorType[FunctionOperatorType["Acos"] = 37] = "Acos";
    FunctionOperatorType[FunctionOperatorType["Asin"] = 38] = "Asin";
    FunctionOperatorType[FunctionOperatorType["Atn2"] = 39] = "Atn2";
    FunctionOperatorType[FunctionOperatorType["BigMul"] = 40] = "BigMul";
    FunctionOperatorType[FunctionOperatorType["Cosh"] = 41] = "Cosh";
    FunctionOperatorType[FunctionOperatorType["Log10"] = 42] = "Log10";
    FunctionOperatorType[FunctionOperatorType["Sinh"] = 43] = "Sinh";
    FunctionOperatorType[FunctionOperatorType["Tanh"] = 44] = "Tanh";
    FunctionOperatorType[FunctionOperatorType["PadLeft"] = 45] = "PadLeft";
    FunctionOperatorType[FunctionOperatorType["PadRight"] = 46] = "PadRight";
    FunctionOperatorType[FunctionOperatorType["StartsWith"] = 47] = "StartsWith";
    FunctionOperatorType[FunctionOperatorType["EndsWith"] = 48] = "EndsWith";
    FunctionOperatorType[FunctionOperatorType["Contains"] = 49] = "Contains";
    FunctionOperatorType[FunctionOperatorType["ToInt"] = 50] = "ToInt";
    FunctionOperatorType[FunctionOperatorType["ToLong"] = 51] = "ToLong";
    FunctionOperatorType[FunctionOperatorType["ToFloat"] = 52] = "ToFloat";
    FunctionOperatorType[FunctionOperatorType["ToDouble"] = 53] = "ToDouble";
    FunctionOperatorType[FunctionOperatorType["ToDecimal"] = 54] = "ToDecimal";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeThisYear"] = 55] = "LocalDateTimeThisYear";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeThisMonth"] = 56] = "LocalDateTimeThisMonth";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeLastWeek"] = 57] = "LocalDateTimeLastWeek";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeThisWeek"] = 58] = "LocalDateTimeThisWeek";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeYesterday"] = 59] = "LocalDateTimeYesterday";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeToday"] = 60] = "LocalDateTimeToday";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeNow"] = 61] = "LocalDateTimeNow";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeTomorrow"] = 62] = "LocalDateTimeTomorrow";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeDayAfterTomorrow"] = 63] = "LocalDateTimeDayAfterTomorrow";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeNextWeek"] = 64] = "LocalDateTimeNextWeek";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeTwoWeeksAway"] = 65] = "LocalDateTimeTwoWeeksAway";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeNextMonth"] = 66] = "LocalDateTimeNextMonth";
    FunctionOperatorType[FunctionOperatorType["LocalDateTimeNextYear"] = 67] = "LocalDateTimeNextYear";
    FunctionOperatorType[FunctionOperatorType["InDateRange"] = 68] = "InDateRange";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalBeyondThisYear"] = 69] = "IsOutlookIntervalBeyondThisYear";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalLaterThisYear"] = 70] = "IsOutlookIntervalLaterThisYear";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalLaterThisMonth"] = 71] = "IsOutlookIntervalLaterThisMonth";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalNextWeek"] = 72] = "IsOutlookIntervalNextWeek";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalLaterThisWeek"] = 73] = "IsOutlookIntervalLaterThisWeek";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalTomorrow"] = 74] = "IsOutlookIntervalTomorrow";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalToday"] = 75] = "IsOutlookIntervalToday";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalYesterday"] = 76] = "IsOutlookIntervalYesterday";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalEarlierThisWeek"] = 77] = "IsOutlookIntervalEarlierThisWeek";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalLastWeek"] = 78] = "IsOutlookIntervalLastWeek";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalEarlierThisMonth"] = 79] = "IsOutlookIntervalEarlierThisMonth";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalEarlierThisYear"] = 80] = "IsOutlookIntervalEarlierThisYear";
    FunctionOperatorType[FunctionOperatorType["IsOutlookIntervalPriorThisYear"] = 81] = "IsOutlookIntervalPriorThisYear";
    FunctionOperatorType[FunctionOperatorType["IsLastMonth"] = 82] = "IsLastMonth";
    FunctionOperatorType[FunctionOperatorType["IsLastYear"] = 83] = "IsLastYear";
    FunctionOperatorType[FunctionOperatorType["IsNextMonth"] = 84] = "IsNextMonth";
    FunctionOperatorType[FunctionOperatorType["IsNextYear"] = 85] = "IsNextYear";
    FunctionOperatorType[FunctionOperatorType["IsThisWeek"] = 86] = "IsThisWeek";
    FunctionOperatorType[FunctionOperatorType["IsThisMonth"] = 87] = "IsThisMonth";
    FunctionOperatorType[FunctionOperatorType["IsThisYear"] = 88] = "IsThisYear";
    FunctionOperatorType[FunctionOperatorType["IsJanuary"] = 89] = "IsJanuary";
    FunctionOperatorType[FunctionOperatorType["IsFebruary"] = 90] = "IsFebruary";
    FunctionOperatorType[FunctionOperatorType["IsMarch"] = 91] = "IsMarch";
    FunctionOperatorType[FunctionOperatorType["IsApril"] = 92] = "IsApril";
    FunctionOperatorType[FunctionOperatorType["IsMay"] = 93] = "IsMay";
    FunctionOperatorType[FunctionOperatorType["IsJune"] = 94] = "IsJune";
    FunctionOperatorType[FunctionOperatorType["IsJuly"] = 95] = "IsJuly";
    FunctionOperatorType[FunctionOperatorType["IsAugust"] = 96] = "IsAugust";
    FunctionOperatorType[FunctionOperatorType["IsSeptember"] = 97] = "IsSeptember";
    FunctionOperatorType[FunctionOperatorType["IsOctober"] = 98] = "IsOctober";
    FunctionOperatorType[FunctionOperatorType["IsNovember"] = 99] = "IsNovember";
    FunctionOperatorType[FunctionOperatorType["IsDecember"] = 100] = "IsDecember";
    FunctionOperatorType[FunctionOperatorType["IsSameDay"] = 101] = "IsSameDay";
    FunctionOperatorType[FunctionOperatorType["IsYearToDate"] = 102] = "IsYearToDate";
    FunctionOperatorType[FunctionOperatorType["DateDiffTick"] = 103] = "DateDiffTick";
    FunctionOperatorType[FunctionOperatorType["DateDiffSecond"] = 104] = "DateDiffSecond";
    FunctionOperatorType[FunctionOperatorType["DateDiffMilliSecond"] = 105] = "DateDiffMilliSecond";
    FunctionOperatorType[FunctionOperatorType["DateDiffMinute"] = 106] = "DateDiffMinute";
    FunctionOperatorType[FunctionOperatorType["DateDiffHour"] = 107] = "DateDiffHour";
    FunctionOperatorType[FunctionOperatorType["DateDiffDay"] = 108] = "DateDiffDay";
    FunctionOperatorType[FunctionOperatorType["DateDiffMonth"] = 109] = "DateDiffMonth";
    FunctionOperatorType[FunctionOperatorType["DateDiffYear"] = 110] = "DateDiffYear";
    FunctionOperatorType[FunctionOperatorType["DateTimeFromParts"] = 111] = "DateTimeFromParts";
    FunctionOperatorType[FunctionOperatorType["GetDate"] = 112] = "GetDate";
    FunctionOperatorType[FunctionOperatorType["GetMilliSecond"] = 113] = "GetMilliSecond";
    FunctionOperatorType[FunctionOperatorType["GetSecond"] = 114] = "GetSecond";
    FunctionOperatorType[FunctionOperatorType["GetMinute"] = 115] = "GetMinute";
    FunctionOperatorType[FunctionOperatorType["GetHour"] = 116] = "GetHour";
    FunctionOperatorType[FunctionOperatorType["GetDay"] = 117] = "GetDay";
    FunctionOperatorType[FunctionOperatorType["GetMonth"] = 118] = "GetMonth";
    FunctionOperatorType[FunctionOperatorType["GetYear"] = 119] = "GetYear";
    FunctionOperatorType[FunctionOperatorType["GetDayOfWeek"] = 120] = "GetDayOfWeek";
    FunctionOperatorType[FunctionOperatorType["GetDayOfYear"] = 121] = "GetDayOfYear";
    FunctionOperatorType[FunctionOperatorType["GetTimeOfDay"] = 122] = "GetTimeOfDay";
    FunctionOperatorType[FunctionOperatorType["Now"] = 123] = "Now";
    FunctionOperatorType[FunctionOperatorType["UtcNow"] = 124] = "UtcNow";
    FunctionOperatorType[FunctionOperatorType["Today"] = 125] = "Today";
    FunctionOperatorType[FunctionOperatorType["AddTimeSpan"] = 126] = "AddTimeSpan";
    FunctionOperatorType[FunctionOperatorType["AddTicks"] = 127] = "AddTicks";
    FunctionOperatorType[FunctionOperatorType["AddMilliSeconds"] = 128] = "AddMilliSeconds";
    FunctionOperatorType[FunctionOperatorType["AddSeconds"] = 129] = "AddSeconds";
    FunctionOperatorType[FunctionOperatorType["AddMinutes"] = 130] = "AddMinutes";
    FunctionOperatorType[FunctionOperatorType["AddHours"] = 131] = "AddHours";
    FunctionOperatorType[FunctionOperatorType["AddDays"] = 132] = "AddDays";
    FunctionOperatorType[FunctionOperatorType["AddMonths"] = 133] = "AddMonths";
    FunctionOperatorType[FunctionOperatorType["AddYears"] = 134] = "AddYears";
    FunctionOperatorType[FunctionOperatorType["OrderDescToken"] = 135] = "OrderDescToken";
})(FunctionOperatorType || (FunctionOperatorType = {}));
