// @devexpress/analytics-core/core/utils/_utils.getResizableOptions.js
/**
* DevExpress Analytics (core\utils\_utils.getResizableOptions.js)
* Version:  24.1.7
* Build date: Oct 28, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { convertFromCssPixelUnits } from '../internal/_utils.unitsConvertation';
import { currentMultiPlatformEngine } from '../../serializer/native/multiplatformEngine';
import { koUtils } from './_koUtils';
import { $dx } from './_jqueryUtils';
export function getResizableOptions($element, panelOffset, minWidth, position, startPosition, width, disabled) {
    const disabledOption = currentMultiPlatformEngine.wrap(koUtils.unwrap(disabled));
    const minWidthOption = koUtils.unwrap(minWidth);
    if (koUtils.isSubscribable(disabled)) {
        disabled.subscribe((newVal) => {
            if (typeof disabledOption === 'function') {
                disabledOption(newVal);
            }
        });
    }
    return {
        starting: function (e) {
            $dx($element).css(position === startPosition ? 'right' : 'left', '');
        },
        handles: currentMultiPlatformEngine.wrap(position === startPosition ? 'e' : 'w'),
        stopped: () => { },
        stop: () => { },
        resize: function (event, element) {
            const startResizePosition = convertFromCssPixelUnits(element.dataset.originalLeftMousePosition);
            const originalWidth = convertFromCssPixelUnits(element.dataset.originalWidth);
            const sizeDiff = event.x - startResizePosition;
            $dx($element).css({ left: position === startPosition ? panelOffset + 'px' : '', right: position === startPosition ? '' : panelOffset + 'px' });
            const newWidth = Math.min(Math.max(minWidthOption, position === startPosition ? originalWidth + sizeDiff : originalWidth - sizeDiff), 1000);
            width && width(newWidth);
        },
        disabled: disabledOption,
        zoom: 1,
        minimumWidth: currentMultiPlatformEngine.wrap(koUtils.unwrap(disabledOption) ? 0 : minWidthOption),
        maximumWidth: 1000,
        $element: $element
    };
}
