// @devexpress/analytics-core/query-builder/dataSource/federation/_dbSchemaFederationDataSourceProvider.js
/**
* DevExpress Analytics (query-builder\dataSource\federation\_dbSchemaFederationDataSourceProvider.js)
* Version:  24.1.7
* Build date: Oct 28, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { DxDeferred } from '../../../core/utils/_jqueryUtils';
import { Disposable } from '../../../serializer/disposable';
import { PathRequest } from '../../../widgets/common/pathRequest';
import { DBTable } from '../dbTable';
export class DBSchemaFederationDataSourceProvider extends Disposable {
    getDbSchema() {
        const deferred = new DxDeferred();
        if (this._rootItems.length)
            deferred.resolve(this._rootItems);
        else {
            this.getItems(new PathRequest('')).done(rootItems => {
                deferred.resolve(rootItems);
            });
        }
        return deferred.promise();
    }
    getDbTable(tableName, path) {
        const $deferred = new DxDeferred();
        this.getItems(new PathRequest(path)).done(items => {
            const columns = items.map(x => {
                return {
                    Name: x.name
                };
            });
            $deferred.resolve(new DBTable({ Name: tableName, columns: columns }));
        });
        return $deferred.promise();
    }
    dispose() {
        super.dispose();
        this.getItems = null;
    }
    constructor(itemsProvider) {
        super();
        this._rootItems = [];
        this.getItems = itemsProvider.getItems;
    }
}
