// devexpress-reporting/designer/controls/crossTab/enums.js
/**
* DevExpress HTML/JS Reporting (designer\controls\crossTab\enums.js)
* Version:  24.1.7
* Build date: Oct 28, 2024
* Copyright (c) 2012 - 2024 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
export var CornerHeaderDisplayMode;
(function (CornerHeaderDisplayMode) {
    CornerHeaderDisplayMode[CornerHeaderDisplayMode["None"] = 0] = "None";
    CornerHeaderDisplayMode[CornerHeaderDisplayMode["RowFieldNames"] = 1] = "RowFieldNames";
    CornerHeaderDisplayMode[CornerHeaderDisplayMode["ColumnFieldNames"] = 2] = "ColumnFieldNames";
})(CornerHeaderDisplayMode || (CornerHeaderDisplayMode = {}));
export var CellKind;
(function (CellKind) {
    CellKind[CellKind["None"] = 0] = "None";
    CellKind[CellKind["Corner"] = 1] = "Corner";
    CellKind[CellKind["RowHeader"] = 2] = "RowHeader";
    CellKind[CellKind["RowTotalHeader"] = 3] = "RowTotalHeader";
    CellKind[CellKind["RowTotal"] = 4] = "RowTotal";
    CellKind[CellKind["ColumnHeader"] = 5] = "ColumnHeader";
    CellKind[CellKind["ColumnTotalHeader"] = 6] = "ColumnTotalHeader";
    CellKind[CellKind["ColumnTotal"] = 7] = "ColumnTotal";
    CellKind[CellKind["Data"] = 8] = "Data";
    CellKind[CellKind["DataHeader"] = 9] = "DataHeader";
    CellKind[CellKind["GrandTotal"] = 10] = "GrandTotal";
    CellKind[CellKind["EmptyHeader"] = 11] = "EmptyHeader";
    CellKind[CellKind["Empty"] = 12] = "Empty";
})(CellKind || (CellKind = {}));
export var TotalHeaderPosition;
(function (TotalHeaderPosition) {
    TotalHeaderPosition[TotalHeaderPosition["Inner"] = 0] = "Inner";
    TotalHeaderPosition[TotalHeaderPosition["Outer"] = 1] = "Outer";
})(TotalHeaderPosition || (TotalHeaderPosition = {}));
export var TotalsPosition;
(function (TotalsPosition) {
    TotalsPosition[TotalsPosition["AfterData"] = 0] = "AfterData";
    TotalsPosition[TotalsPosition["BeforeData"] = 1] = "BeforeData";
})(TotalsPosition || (TotalsPosition = {}));
export var DataFieldLayout;
(function (DataFieldLayout) {
    DataFieldLayout[DataFieldLayout["InRow"] = 0] = "InRow";
    DataFieldLayout[DataFieldLayout["InColumn"] = 1] = "InColumn";
})(DataFieldLayout || (DataFieldLayout = {}));
